streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -200,6 +201,7 @@ synchronized void evict() {
         }
         remove(eldest);
         cache.remove(eldest.key);
+        dirtyKeys.remove(eldest.key);
```

Comment:
It's technically correct, but we must add a comment here, otherwise people will wonder why isn't flush() sufficient. Actually I'm still not sure why flush is not sufficient under normal (non-test) circumstances, since I can't see how you'd have a clean entry overwriting a dirty entry. The clean entry is inserted in the cache during get(). If the dirty entry was there, why didn't get() pick it up? Again, this is a broader comment, I think the PR should still go through with this, but there might be additional issues.